tests/test_agents/base.py

```diff
@@ -434,53 +413,75 @@ class BaseTestEnd2EndAgentCatchup(BaseTestEnd2End):
+            )
+        if self.n_terminal and not hasattr(self, "stop_string"):
+            pytest.fail("'stop_string' must be provided for agent termination.")
 
     def test_run(self, nb_nodes: int) -> None:
         """Run the test."""
+        self.nb_nodes = nb_nodes
+        self.check()
+
         self.prepare_and_launch(nb_nodes)
         self.health_check(
             max_retries=self.HEALTH_CHECK_MAX_RETRIES,
             sleep_interval=self.HEALTH_CHECK_SLEEP_INTERVAL,
         )
-        self._stop_and_restart_last_agent(nb_nodes)
+        if self.n_terminal:
+            self._restart_agents()
         self.check_aea_messages()
+        self.terminate_processes()
+
+    def _restart_agents(self) -> None:
+        """Stops and restarts agents after stop string is found."""
 
-    def _stop_and_restart_last_agent(self, nb_agents: int) -> None:
-        """Stops and restarts the last agents when stop string is found."""
         # stop the last agent as soon as the "stop string" is found in the output
+        # once found, we start terminating the first agent
         logging.info(f"Waiting for string {self.stop_string} in last agent output")
         missing_strict_strings, _ = self.missing_from_output(
-            process=self.processes[-1],
+            process=self.processes[max(self.processes)],
             strict_check_strings=(self.stop_string,),
             timeout=self.wait_before_stop,
         )
         if missing_strict_strings:
-            raise RuntimeError("cannot stop agent correctly")
-        logging.info("Last agent stopped")
+            msg = f"cannot stop agent, stop string `{self.stop_string}` not found"
+            raise RuntimeError(msg)
 
-        # immediately terminate the agent
-        self.terminate_agents(self.processes[-1], timeout=0)
+        if self.wait_to_kill:
+            logging.info("Waiting to terminate agents")
+            time.sleep(self.wait_to_kill)
+
+        # terminate the agents - sequentially
         # don't pop before termination, seems to lead to failure!
-        self.processes.pop()
+        for i in range(self.n_terminal):
+            agent_name = self.__get_agent_name(i)
+            self.terminate_agents(self.processes[i], timeout=0)
+            self.processes.pop(i)
+            logging.info(f"Terminated {agent_name}")
 
         # wait for some time before restarting
         logging.info(
             f"Waiting {self.restart_after} seconds before restarting the agent"
         )
         time.sleep(self.restart_after)
 
-        # restart agent
-        logging.info("Restart the agent")
-        self._launch_agent_i(-1, nb_agents)
-        self.check_aea_messages()
-        self.terminate_processes()
+        # restart agents
+        for i in range(self.n_terminal):
+            agent_name = self.__get_agent_name(i)
+            self._launch_agent_i(i)
+            logging.info(f"Restarted {agent_name}")
```

Comment:
here we now also iterate from 0. Since we use a mapping the "order" of the processes is guaranteed to remain correct.